src/libsyntax/parse/parser.rs

```diff
@@ -2870,7 +2870,46 @@ impl<'a> Parser<'a> {
                 let (span, e) = self.interpolated_or_expr_span(e)?;
                 (lo.to(span), ExprKind::Box(e))
             }
-            _ => return self.parse_dot_or_call_expr(Some(attrs))
+            token::Ident(..) if self.token.is_ident_named("not") => {
+                // `not` is just an ordinary identifier in Rust-the-language,
+                // but as `rustc`-the-compiler, we can issue clever diagnostics
+                // for confused users who really want to say `!`
+                let token_cannot_continue_expr = |t: &token::Token| match *t {
+                    // These tokens can't continue an expression after an ident
+                    token::Ident(..) | token::Literal(..) | token::Pound => true,
+                    _ => false
+                };
+                let cannot_continue_expr = self.look_ahead(1, token_cannot_continue_expr);
+                if cannot_continue_expr {
+                    self.bump();
+                    let found = match self.token {
+                        // FIXME: map keywords to `TokenType::Keyword`
+                        token::Ident(..) => TokenType::Ident,
+                        _ => TokenType::Token(self.token.clone())
+                    };
+                    // Emit the error ...
+                    let mut err = self.diagnostic()
+                        .struct_span_err(self.span,
+                                         &format!("unexpected {} after identifier",
+                                                  found.to_string()));
```

Comment:
You can use `format!("unexpected {} after identifier", self.this_token_descr())` here.